OrbitGl/GlSlider.cpp

```diff
@@ -83,99 +148,118 @@ void GlSlider::DrawSlider(GlCanvas* canvas, float x, float y, float width, float
   const Color dark_border_color = GetDarkerColor(bar_color_);
 
   // Background
-  if (!picking) {
-    DrawBackground(canvas, x, GetOrthogonalSliderSize(), GetPixelHeight(), canvasHeight);
-  }
-
-  float start = (1.0f - ratio_) * nonSliderHeight + GetOrthogonalSliderSize();
-
-  DrawSlider(canvas, x, start, GetPixelHeight(), sliderHeight, ShadingDirection::kRightToLeft);
-}
+  DrawBackground(canvas, x, GetOrthogonalSliderSize(), GetPixelHeight(), canvas_height);
 
-void GlHorizontalSlider::OnPick(int x, int /*y*/) {
-  float canvasWidth = canvas_->GetWidth();
-  float sliderWidth = length_ * canvasWidth;
-  float nonSliderWidth = canvasWidth - sliderWidth;
+  float start = ceilf((1.0f - pos_ratio_) * non_slider_height + GetOrthogonalSliderSize());
 
-  float sliderX = ratio_ * nonSliderWidth;
-  picking_ratio_ = (static_cast<float>(x) - sliderX) / sliderWidth;
+  DrawSlider(canvas, x, start, GetPixelHeight(), slider_height, ShadingDirection::kRightToLeft);
 }
 
-void GlHorizontalSlider::OnDrag(int x, int /*y*/) {
-  float canvasWidth = canvas_->GetWidth();
-  float sliderWidth = length_ * canvasWidth;
-  float nonSliderWidth = canvasWidth - sliderWidth;
-
-  float sliderLeftWidth = picking_ratio_ * length_ * canvasWidth;
-  float newX = static_cast<float>(x) - sliderLeftWidth;
-  float ratio = newX / nonSliderWidth;
-
-  ratio_ = clamp(ratio, 0.f, 1.f);
-
-  if (drag_callback_) {
-    drag_callback_(ratio_);
-  }
+float GlVerticalSlider::GetCanvasEdgeLength() {
+  return canvas_->GetHeight() - GetOrthogonalSliderSize();
 }
 
-void GlHorizontalSlider::Draw(GlCanvas* canvas, PickingMode picking_mode) {
-  const bool picking = picking_mode != PickingMode::kNone;
-
+void GlHorizontalSlider::Draw(GlCanvas* canvas, PickingMode /*picking_mode*/) {
   canvas_ = canvas;
   static float y = 0;
 
-  float canvasWidth = canvas_->GetWidth() - GetOrthogonalSliderSize();
-  float sliderWidth = length_ * canvasWidth;
-  float nonSliderWidth = canvasWidth - sliderWidth;
+  float canvas_width = canvas_->GetWidth() - GetOrthogonalSliderSize();
+  float slider_width = ceilf(length_ratio_ * canvas_width);
+  float non_slider_width = canvas_width - slider_width;
 
   const Color dark_border_color = GetDarkerColor(bar_color_);
 
   Color color =
       canvas->GetPickingManager().IsThisElementPicked(this) ? selected_color_ : slider_color_;
   const Color light_border_color = GetLighterColor(color);
 
-  if (!picking) {
-    DrawBackground(canvas, 0, y, canvasWidth, GetPixelHeight());
-  }
+  DrawBackground(canvas, 0, y, canvas_width, GetPixelHeight());
+
+  float start = floorf(pos_ratio_ * non_slider_width);
+
+  DrawSlider(canvas, start, y, slider_width, GetPixelHeight(), ShadingDirection::kTopToBottom);
+
+  Batcher* batcher = canvas->GetBatcher();
+  const float kEpsilon = 0.0001f;
 
-  float start = ratio_ * nonSliderWidth;
+  // Left / right resize arrows and separator
+  const float height_factor = 2.f;
+  float tri_size = slider_resize_pixel_margin_ - 3.f;
```

Comment:
The slider has a 2px border, and since the available space for the arrow is quite limited, I've decided to have the tip of the arrow touch this border and leave 1px between the vertical line and the arrow. It's a bit arbitrary, but IMHO it looks better than a super-small arrow.

I'll add a comment to explain that.